api/python/tests/integration/test_packages.py

```diff
@@ -666,6 +666,61 @@ def test_s3_set_dir(self):
 
             list_object_versions_mock.assert_called_with('bucket', 'foo/')
 
+    def test_set_dir_update_policy(self):
+        """Verify building a package with update policy. """
+
+        # Test for local src dir
+        foodir = pathlib.Path("foo_dir")
+        bazdir = pathlib.Path("baz_dir")
+        foodir.mkdir()
+        bazdir.mkdir()
+        with open(foodir / 'bar', 'w') as fd:
+            fd.write(fd.name)
+        with open(bazdir / 'bar', 'w') as fd:
+            fd.write(''.join([str(i) for i in range(100)]))
+
+        pkg = Package()
+        pkg = pkg.set_dir("/", ".", meta={'name': 'test_meta'})
+        assert 'foo_dir' in pkg.keys()
+        existing_bar_size = pkg['foo_dir/bar'].size
+
+        # existing update policy
+        pkg = pkg.set_dir("foo_dir", bazdir, update_policy='existing')
+        assert 'foo_dir' in pkg.keys()
+        assert existing_bar_size == pkg['foo_dir/bar'].size
+
+        # incoming update policy
+        pkg = pkg.set_dir("foo_dir", bazdir)
+        assert 'foo_dir' in pkg.keys()
+        assert existing_bar_size != pkg['foo_dir/bar'].size
+
+        # verify non existing update policy raises value error
+        with pytest.raises(ValueError):
+            pkg.set_dir("foo_dir", foodir, update_policy='invalid_policy')
+
+        # Test for s3 src dir
+        with patch('quilt3.packages.list_object_versions') as list_object_versions_mock:
+            pkg = Package()
+
+            list_object_versions_mock.return_value = ([
```

Comment:
Might as well return more than one object since this is a set_dir call.

Reply:
If you haven't already, do a reall list_object_versions call and check your mock against that. Try to match all fields.

Reply:
Since you are mocking an S3 call, we should be mocking this with boto3, so use `self.s3_stubber.add_response`

Reply:
I'm replicating what we already have for testing [s3_set_dir](https://github.com/quiltdata/quilt/blob/bd1a3fa69f3a3332e69d7ba46b16692497cbeb0c/api/python/tests/integration/test_packages.py#L504). This time just adding update policy

Reply:
That's not really correct on a few levels: first because we should be using boto3, and second because when you have the same test with small variations we should probably be using [parameterize](https://github.com/quiltdata/quilt/blob/master/lambdas/thumbnail/tests/test_thumbnail.py#L37) or similar. Have a look. I think it's worth rewriting the test for `set_dir` entirely and doing it with parameterize. But you'll have to feel it out by trying it.

Reply:
Actually, skip parameterize and rewriting the set_dir test for now, but definitely use boto3 in your new test. If that fails you can do it the way Kevin did in `set_dir`.

Reply:
Looking at this again, I think it's best to leave it as it is (the way Kevin did in `set_dir`). 
Reason being that `list_object_versions_mock ` is for `quilt3.packages.list_object_versions` which is a [recursive method](https://github.com/quiltdata/quilt/blob/4d03796b433e693443a2f535dfe526211d6fff17/api/python/quilt3/data_transfer.py#L581) in data_trannsfer. So it doesn't have any direct call with s3 boto call.